python/ray/_private/event/export_event_logger.py

```diff
@@ -0,0 +1,137 @@
+        self.logger.handlers[0].flush()
+
+    def _create_export_event(self, event_data: ExportEventDataType) -> ExportEvent:
+        event = ExportEvent()
+        event.event_id = get_event_id()
+        event.timestamp = int(datetime.now().timestamp())
+        if type(event_data) is ExportSubmissionJobEventData:
+            event.submission_job_event_data.CopyFrom(event_data)
+            event.source_type = ExportEvent.SourceType.EXPORT_SUBMISSION_JOB
+        else:
+            raise TypeError(f"Invalid event_data type: {type(event_data)}")
+        if event.source_type != self.source:
+            global_logger.error(
+                f"event_data has source type {event.source_type}, however "
+                f"the event was sent to a logger with source {self.source}. "
+                f"The event will still be written to the file of {self.source} "
+                "but this indicates a bug in the code."
+            )
+            pass
+        return event
+
+    def _export_event_to_string(self, event: ExportEvent) -> str:
+        event_data_json = {}
+        proto_to_dict_options = {
+            "always_print_fields_with_no_presence": True,
+            "preserving_proto_field_name": True,
+            "use_integers_for_enums": False,
+        }
+        event_data_field_set = event.WhichOneof("event_data")
+        if event_data_field_set:
+            event_data_json = message_to_dict(
+                getattr(event, event_data_field_set),
+                **proto_to_dict_options,
+            )
+        else:
+            global_logger.error(
+                f"event_data missing from export event with id {event.event_id} "
+                f"and type {event.source_type}. An empty event will be written, "
+                "but this indicates a bug in the code. "
+            )
+            pass
+        event_json = {
+            "event_id": event.event_id,
+            "timestamp": event.timestamp,
+            "source_type": ExportEvent.SourceType.Name(event.source_type),
+            "event_data": event_data_json,
+        }
+        return json.dumps(event_json)
+
+
+def _build_export_event_file_logger(source: ExportEvent.SourceType, sink_dir: str):
+    logger = logging.getLogger("_ray_export_event_logger")
+    logger.setLevel(logging.INFO)
+    dir_path = pathlib.Path(sink_dir) / "events"
+    filepath = dir_path / f"event_{source}.log"
+    dir_path.mkdir(exist_ok=True)
+    filepath.touch(exist_ok=True)
+    # Configure the logger.
+    handler = logging.handlers.RotatingFileHandler(
+        filepath, maxBytes=(100 * 1e6), backupCount=20  # 100 MB max file size
+    )
+    logger.addHandler(handler)
+    logger.propagate = False
+    return logger
+
+
+# This lock must be used when accessing or updating global event logger dict.
+_export_event_logger_lock = threading.Lock()
+_export_event_logger = {}
+
+
+def get_export_event_logger(source: ExportEvent.SourceType, sink_dir: str):
```

Comment:
can you check how we do structured logging in python layer and see if we should follow similar patterns? 

Do they also use a singleton pattern like this?

Reply:
Structured logging fetches various registered loggers for the different source types, but the existing python events use this singleton pattern. We want to use a single logger to avoid any concurrency issues from multiple threads. Using register logger is an option which also is a singleton, but I think this API is cleaner because there is no separate initialization and get functions.